### ISSUE TITLE
Some PNG transparent images have rendering issues on iOS.

### ISSUE THREAD

**Issue by miluduo** [flutter/flutter#151210]
### Steps to reproduce

This image has rendering issues on iOS.
![test](https://github.com/flutter/flutter/assets/16186425/e8f3617a-c648-4860-81f3-d048096ecfa8)

### Expected results

It displays well on Android
![WechatIMG40](https://github.com/flutter/flutter/assets/16186425/ecf105be-4d58-4fa3-bd80-f4983a02b5e7)

### Actual results

but not on iOS.
![IMG_5930](https://github.com/flutter/flutter/assets/16186425/9884ef31-cc28-4db8-b1e6-d10960baf238)

### Code sample

<details open><summary>Code sample</summary>

```dart
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.black,
      body: Center(child: Image.asset('assets/images/test.png')),
    );
  }
```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 3.22.2, on macOS 14.2.1 23C71 darwin-arm64, locale
    zh-Hans-CN)
[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
[✓] Xcode - develop for iOS and macOS (Xcode 15.0.1)
[✓] Chrome - develop for the web
[✓] Android Studio (version 2023.2)
[✓] VS Code (version unknown)
    ✗ Unable to determine VS Code version.
[✓] VS Code (version unknown)
    ✗ Unable to determine VS Code version.
[✓] Proxy Configuration
[✓] Connected device (5 available)
[✓] Network resources

• No issues found!
```

</details>

**Comment by danagbemava-nc** [comment]
Reproducible using the image provided. This appears to be an issue an with the impeller rendering engine. 

This is also more visible in dark mode as opposed to light mode. 

| skia | impeller | 
| -- | -- |
| <img src="https://github.com/flutter/flutter/assets/88313112/e538d2e6-be0e-4cfe-9928-6a65c2ff7264" width="300" /> | <img src="https://github.com/flutter/flutter/assets/88313112/22332f0e-7cbe-40dc-9d56-b86e2337e54c" width="300" /> |

<details>
<summary>full sample</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
        brightness: Brightness.dark
      ),
      home: Scaffold(
        body: Center(
          child: Image.asset("assets/images/image.PNG")
        ),
      ),
    );
  }
}
```

</details>

<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel stable, 3.22.2, on macOS 14.4.1 23E224 darwin-arm64, locale en-GB)
    • Flutter version 3.22.2 on channel stable at /Users/nexus/dev/sdks/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 761747bfc5 (4 weeks ago), 2024-06-05 22:15:13 +0200
    • Engine revision edd8546116
    • Dart version 3.4.3
    • DevTools version 2.34.3

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/nexus/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • Java binary at: /Users/nexus/Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.10+0-17.0.10b1087.21-11609105)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.3)
    • Xcode at /Applications/Xcode-15.3.0.app/Contents/Developer
    • Build 15E204a
    • CocoaPods version 1.15.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2024.1)
    • Android Studio at /Users/nexus/Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.10+0-17.0.10b1087.21-11609105)

[✓] IntelliJ IDEA Ultimate Edition (version 2023.2.5)
    • IntelliJ at /Users/nexus/Applications/IntelliJ IDEA Ultimate.app
    • Flutter plugin version 77.2.2
    • Dart plugin version 232.10286

[✓] VS Code (version 1.90.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.90.0

[✓] Connected device (5 available)
    • Nexus (mobile)                  • 00008020-001875E83A38002E • ios            • iOS 17.5.1 21F90
    • Dean’s iPad (mobile)            • 00008103-000825C811E3401E • ios            • iOS 17.5.1 21F90
    • macOS (desktop)                 • macos                     • darwin-arm64   • macOS 14.4.1 23E224 darwin-arm64
    • Mac Designed for iPad (desktop) • mac-designed-for-ipad     • darwin         • macOS 14.4.1 23E224 darwin-arm64
    • Chrome (web)                    • chrome                    • web-javascript • Google Chrome 126.0.6478.127
    ! Error: Dean’s Apple Watch has not finished loading development services. Please select a different device, or wait for the device to load development services and try again. (code 8)

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```
  
```
[!] Flutter (Channel master, 3.23.0-14.0.pre.117, on macOS 14.4.1 23E224 darwin-arm64, locale en-GB)
    • Flutter version 3.23.0-14.0.pre.117 on channel master at /Users/nexus/dev/sdks/flutters
    ! Warning: `flutter` on your path resolves to /Users/nexus/dev/sdks/flutter/bin/flutter, which is not inside your current Flutter SDK checkout at /Users/nexus/dev/sdks/flutters. Consider adding /Users/nexus/dev/sdks/flutters/bin to the front of your path.
    ! Warning: `dart` on your path resolves to /Users/nexus/dev/sdks/flutter/bin/dart, which is not inside your current Flutter SDK checkout at /Users/nexus/dev/sdks/flutters. Consider adding /Users/nexus/dev/sdks/flutters/bin to the front of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision f20962672e (6 hours ago), 2024-07-02 22:59:23 -0400
    • Engine revision c5c0c54d6d
    • Dart version 3.5.0 (build 3.5.0-323.0.dev)
    • DevTools version 2.37.0
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform update checks and upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/nexus/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • Java binary at: /Users/nexus/Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.10+0-17.0.10b1087.21-11609105)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.3)
    • Xcode at /Applications/Xcode-15.3.0.app/Contents/Developer
    • Build 15E204a
    • CocoaPods version 1.15.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2024.1)
    • Android Studio at /Users/nexus/Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.10+0-17.0.10b1087.21-11609105)

[✓] IntelliJ IDEA Ultimate Edition (version 2023.2.5)
    • IntelliJ at /Users/nexus/Applications/IntelliJ IDEA Ultimate.app
    • Flutter plugin version 77.2.2
    • Dart plugin version 232.10286

[✓] VS Code (version 1.90.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.90.0

[✓] Connected device (5 available)
    • Nexus (mobile)                  • 00008020-001875E83A38002E • ios            • iOS 17.5.1 21F90
    • Dean’s iPad (mobile)            • 00008103-000825C811E3401E • ios            • iOS 17.5.1 21F90
    • macOS (desktop)                 • macos                     • darwin-arm64   • macOS 14.4.1 23E224 darwin-arm64
    • Mac Designed for iPad (desktop) • mac-designed-for-ipad     • darwin         • macOS 14.4.1 23E224 darwin-arm64
    • Chrome (web)                    • chrome                    • web-javascript • Google Chrome 126.0.6478.127
    ! Error: Dean’s Apple Watch has not finished loading development services. Please select a different device, or wait for the device to load development services and try again. (code 8)

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.
```
</details>

**Comment by jonahwilliams** [comment]
Difference with impeller is that we're decoding to a 64 bit color format to support wide gamut. We could quickly check if disabling wide gamut fixes the rendering bug.

**Comment by jtmcdole** [comment]
This is not wide gamut and instead unpremultiplied image alpha. Working on a fix now; but as a work around you could:

a: premultiply your alpha
b: disable Impeller and switch back to skia

**Comment by jtmcdole** [comment]
Producing a test image was "fun" - turns out the problem is _probably_ somewhere upstream in the APNG code base. I wasn't able to reproduce the problem still I noticed all my images were missing the `acTL` header; upon adding that (test image is now a green dot surrounded by unmultiplied red) I was able to reproduce the problem.  The fix in flutter/engine#53770 is still good, but it would be great to track down any upstream issues as well.
